Issues in getConnectors and a databaseconnection field

elinw opened the pull request:
This addresses some problems in getConnectors which made it not work with the current file structure in the database folder and with the
newer naming conventions for the database class. 
It also adds a new field that uses getConnectors to dynamically create a
list of available database connection/driver combinations optionally
constrained to a list of supported databases for the application.

This will be important for multidatabase support since not all servers
will support all databases and not all applications will support all
available drivers.

Commit:
Only setting the viewLayout if a layout is passed in the request. The default value should not be set here as it already has been set by the JView constructor, allowing for a view to have a different default value than 'default'.

elinw committed:
This addresses some problems in getConnectors which made it not work
with the current file structure in the database folder and with the
newer naming conventions for the database class. 
It also adds a new field that uses getConnectors to dynamically create a
list of available database connection/driver combinations optionally
constrained to a list of supported databases for the application.

This will be important for multidatabase support since not all servers
will support all databases and not all applications will support all
available drivers.

chdemko commented on libraries/joomla/database/database.php:
```diff
@@ -169,57 +177,71 @@ abstract class JDatabase implements JDatabaseInterface
 
 		// Get a list of types.
-		$types = JFolder::folders(dirname(__FILE__));
+		$types = JFolder::files(dirname(__FILE__).'/database');
```
you can add some parameters to JFolder::files to filters the files you need (avoiding the tests done after)

elinw replied:
I'm not sure what parameters you might mean given that we are trying to pick up the databases dynamically from the file names.  The point being we don't know their names in advance.

chdemko replied:
after that line, you are excluding some files (containing 'exporter, 'importer', ...). Yo can do that while you are calling JFolder::files

elinw replied:
It is better to do it that way than writing regex for those strings.

chdemko commented on libraries/joomla/database/database.php:
```diff
@@ -169,57 +177,71 @@ abstract class JDatabase implements JDatabaseInterface
+					// If the file exists register the class with our class loader.
+					if (file_exists($path))
+					{
+						JLoader::register($class, $path);
```
I think you have to get a third parameter to JLoader::register: true to force the update of the classes list in the loader

elinw replied:
No, I don't think that is correct. We are just making a list here, not overriding.

chdemko commented on libraries/joomla/database/database.php:
```diff
@@ -169,57 +177,71 @@ abstract class JDatabase implements JDatabaseInterface
+					$connectors[] = $type;
+				}
+			}		
+			self:$connectors = $connectors;
```
use self::$connectors (not self:$connectors)

chdemko commented on libraries/joomla/database/database.php:
```diff
@@ -169,57 +177,71 @@ abstract class JDatabase implements JDatabaseInterface
+			self:$connectors = $connectors;
+			
+			return self:$connectors;
 	}
```
it seems to lack a closing bracket

chdemko commented:
Would it be possible to put all files concerning a connector in a folder named

libraries/joomla/database/database/_connector_

i.e. for mysql:

libraries/joomla/database/database/mysql

and name these files

libraries/joomla/database/database/mysql/driver.php: define class JDatabaseMysqlDriver
libraries/joomla/database/database/mysql/exporter.php: define class JDatabaseMysqlExporter
libraries/joomla/database/database/mysql/importer.php: define class JDatabaseMysqlImporter
libraries/joomla/database/database/mysql/query.php: define class JDatabaseMysqlQuery

when a JDatabase::getInstance or JDatabase::getConnectors is called, do a JLoader::discover('JDatabase'.$type, dirname(__FILE__) . '/' . $type);

and replace
$class = 'JDatabaseDriver' . ucfirst(trim($type));
by
$class = 'JDatabase'.ucfirst(trim($type)).'Driver';

elinw commented:
Christophe, 

That idea was discussed several months ago and will happen eventually. As you can tell, the code I submitted alreay begins the process of looking in folders with the same names as the types.  Please note that this is not a refactoring this is a bug fix (getConnectors is looking in the wrong places for files, for the wrong class names, and so on) and attempt to add a field that is absolutely necessary to be in 11.3 if we are going to have useful multiple database support in any applications.  The bug fixes to getConnector are only consequent to making the field. In the future someone can take up refactoring including reorganizing the files but that is not what this is about.  Let's try to avoid scope creep on bug fixes.

elinw committed:
Add test infrastructure for databaseconnection field.

elinw committed:
Small fixes to getConnectors

elinw commented:
Ok fixed the missing : and } .

elinw committed:
Another :

Commit:
Making the improvements suggested by Hackwar

LouisLandry commented:
So I think if you look at the current version of this method in staging: https://github.com/joomla/joomla-platform/blob/staging/libraries/joomla/database/database.php#L161

There are really only three lines that need to change here: 176, 188 and 195.  Every other aspect of this method should work very well as expected.  The form field I am fine with, but as for the JDatabase::getConnectors() method I don't think it requires memory caching or that much change.

elinw commented:
Should we look for driver.php? It seems out of bounds to me at this point. 

In other words are you saying just do $type.php and be done?

Commit:
Modified behaviour of fetchConfigurationData to improve flexibility and avoid conflicts within unit tests.

Commit:
Add test for JHtmlBatch::language

Commit:
Rename for reuse

Commit:
Modify JHtml helper test cases

Correct path to JHtmlTest.xml in JHtmlBatchTest, add test case for JHtmlList::users

Commit:
Add nouser param

Commit:
Add second test case to properly test nouser

Commit:
Add test cases for JHtmlTel

Commit:
Add test case for JHtmlList::positions

Commit:
Docblocks, remove unneeded setup, comments

Commit:
Add JHtmlMenu test cases

Commit:
Rename JHtmlNumber.php to JHtmlNumberTest.php

Commit:
Add test case for empty bytes

Commit:
Add JHtmlEmail test cases

eddieajau committed:
Merge pull request #437 from mbabker/tests

Improve JHtml test coverage

chdemko committed:
Merge pull request #416 from Bakual/staging

Default Layout should not be set in JController->display() as it's already been set in the JView Constructor

LouisLandry committed:
Merge pull request #440 from ianmacl/fetchconfigdata

Changes to fetchConfigData to increase flexibility and avoid testing conflicts with multiple JConfig classes.

Commit:
Fixed style errors

chdemko committed:
Fix typo

chdemko committed:
Merge pull request #441 from ianmacl/stylefixes

Fixed improperly aligned doc blocks in cli.php and web.php

eddieajau committed:
Merge pull request #442 from chdemko/daemon

Fix typo in daemon.php

eddieajau committed:
Merge pull request #446 from rvsjoen/issue_443

JHtmlNumber and tests fail on 32 bit systems

elinw committed:
Merge branch 'getconnectors' of https://github.com/elinw/joomla-platform into getconnectors

elinw commented:
sigh i's going to be easier to close this.
